// @s3dking/randomjoke/index.js
const dadJokes = [
    "Why don't eggs tell jokes? They'd crack each other up!",
    "What do you call a bear with no teeth? A gummy bear!",
    "Why don't melons get married? They cantaloupe!",
    "What do you call a fake noodle? An impasta!",
    "Why did the scarecrow win an award? He was outstanding in his field!",
    "Why did the math book look sad? Because it had too many problems!",
    "What did the ocean say to the shore? Nothing, it just waved!",
    "Why don't scientists trust atoms? Because they make up everything!",
    "What did the grape say when it got stepped on? Nothing, it just let out a little wine!",
    "Why did the golfer bring two pairs of pants? In case he got a hole in one!",
    "What do you call a bear with no ears? B!",
    "What did the coffee report to the police? A mugging!",
    "Why don't oysters donate to charity? Because they're shellfish!",
    "What did one wall say to the other wall? I'll meet you at the corner!",
    "Why did the bicycle fall over? Because it was two-tired!",
    "What do you call cheese that isn't yours? Nacho cheese!",
    "Why did the tomato turn red? Because it saw the salad dressing!",
    "What do you call a snowman with a six-pack? An abdominal snowman!",
    "Why did the chicken join a band? Because it had the drumsticks!",
    "What do you call a fish with no eyes? Fsh!",
    "Why don't skeletons fight each other? They don't have the guts!",
    "What do you call a can opener that doesn't work? A can't opener!",
    "Why did the golfer bring an extra pair of socks? In case he got a hole in one!",
    "What do you call a factory that makes good products? A satisfactory!",
    "Why did the scarecrow become a successful neurosurgeon? He was outstanding in his field!",
    "What do you call a pig that does karate? A pork chop!",
    "What do you call a dog magician? A labracadabrador!",
    "Why did the cookie go to the doctor? Because it was feeling crumbly!",
    "What did the baby corn say to the mama corn? Where's pop corn?",
    "What do you call a sleeping bull? A bulldozer!",
    "What kind of tree fits in your hand? A palm tree!",
    "What did the buffalo say to his son when he left? Bison!",
    "What did the clock do when it was hungry? It went back four seconds!",
    "What do you call a bee that can't make up its mind? A maybe!",
    "Why did the gym close down? It just didn't work out!",
    "What do you call a fake stone in Ireland? A sham rock!",
    "Why don't pirates take a shower before they walk the plank? They'll just wash up on shore later!",
    "What did the janitor say when he jumped out of the closet? Supplies!",
    "What do you call a duck that gets all A's? A wise quacker!",
    "Why did the jellybean go to school? To become a smartie!",
    "What did the grape do when he got stepped on? He let out a little wine!",
    "What did one hat say to the other? You stay here, I'll go on ahead!",
    "Why did the belt go to jail? For holding up pants!",
    "What do you call a dinosaur that crashes his car? Tyrannosaurus wrecks!",
    "What do you call a pile of cats? A meow-ntain!",
    "Why don't bikes work when they're tired? Because they're two-tired!",
    "What did the paper say to the pencil? Write on!",
    "What do you call a sleeping pizza? A pizzzzza!",
    "What did one elevator say to the other? I think I'm coming down with something!",
    "Why did the invisible man turn down the job offer? He couldn't see himself doing it!",
    "Why did the orange stop rolling down the hill? It ran out of juice!"
];

const darkJokes = [
    "I asked my wife to let me know the next time she has an orgasm. She said she doesn’t like to bother me when I’m at work.",
    "I have a fear of getting buried alive, but I’m slowly working through it.",
    "I started a new job as a human cannonball. My first day was pretty explosive.",
    "I told my friend 10 jokes to make him laugh. Sadly, no pun in ten did.",
    "Why don’t graveyards have 4G? Because they’re full of dead zones.",
    "I used to be a mortician, but I couldn’t deal with the pressure.",
    "I’m reading a book about anti-gravity. It’s impossible to put down. Kind of like my last relationship.",
    "I wanted to be a comedian, but I realized my humor wasn’t the only thing that was dead inside.",
    "The doctor told me I was overweight. I said I want a second opinion. He said, ‘You’re ugly too.’",
    "I met a girl with a tattoo of a tombstone. I thought it was a sign.",
    "I had a job as a banker, but I lost interest.",
    "I’m trying to lose weight, but I keep eating my feelings. Now I’m feeling a lot of regret.",
    "I can’t believe I got fired from the calendar factory. All I did was take a day off.",
    "I have a lot of dark humor, but I’m really just trying to keep the lights on.",
    "My therapist says I have a fear of commitment. I’m not sure if I’ll go back next week.",
    "I used to play piano by ear, but now I use my hands, because the last time I tried it… well, you know.",
    "They say money can’t buy happiness, but it sure can buy a lot of distractions from the fact you’re miserable.",
    "I tried to make a belt out of watches, but it was a waist of time.",
    "Why don’t skeletons fight each other? They don’t have the guts… but they sure love a good bone to pick.",
    "My friend says he’s afraid of commitment. I told him to get a job as a cashier at the graveyard.",
    "I’ve been meaning to get a therapist, but then I remembered I’m too messed up for even them to fix.",
    "I tried to join a support group, but I was told I wasn’t broken enough to fit in.",
    "I’m starting a new diet. It’s called 'Eat Until You’re Dead.'",
    "I once went to a haunted house. I was more terrified by the prices of the tickets.",
    "I don’t need therapy. I just need someone to hear my problems and then forget about them.",
    "I tried to sell my soul, but no one wanted it. Apparently, it's been used up.",
    "I’m on a whiskey diet. I’ve lost three days already.",
    "I walked into a bar and told them, 'I’m here for the open mic.' They said, 'We don’t have open mic nights.' I said, 'Then I’ll just drink till I can’t remember why I came.'",
    "I don’t trust stairs. They’re always up to something... and sometimes, they just let me fall.",
    "I thought I had a fear of death, but turns out I just don’t care anymore.",
    "I’ve started investing in stocks: beef, chicken, and vegetable. One day I hope to be a bouillonaire.",
    "I once asked a friend if they’d help me bury a body. They said, 'That depends... can we start with yours?'",
    "I’ve always wanted to be a cannibal, but I’ve got cold feet about it.",
    "The best part about being alone is that no one’s around to see you cry about it.",
    "I tried to be a baker, but I kneaded dough. Now I'm just kneading help.",
    "I bought a mirror the other day, but it didn’t reflect well on me.",
    "I used to be a pro at dodgeball... until my own thoughts started throwing things at me.",
    "I got caught sneaking into a funeral home. I guess you could say I was just looking for a little 'rest.'",
    "I had a nightmare that I was a muffler. I woke up exhausted, but then I realized it was just the sound of my life going down the drain.",
    "I used to be an optimist, but now I'm just trying to survive each day.",
    "I told my wife she was drawing her eyebrows too high. She looked surprised... and now I’m wondering if I’ll ever wake up.",
    "I’ve been working on my obituary. I want it to be more exciting than my actual life.",
    "I told my kids I was going to start a diet. They looked at me and said, 'Is it too late to start planning your funeral?'",
    "I’m not saying life is a joke, but my therapist is getting paid to laugh at it.",
    "I got a job at a bakery because I kneaded dough... but turns out they didn’t need me around for long.",
    "I wanted to learn how to swim, but I realized that I drown better in misery.",
    "I’m not a fan of death, but I’m starting to understand the appeal.",
    "Why do vampires always seem sick? They’re always coffin.",
    "I’m trying to lose weight, but it’s hard when I have nothing to live for.",
    "I once tried to start a band called 'The Grim Reapers,' but we couldn’t get anyone to 'show up' to the gigs.",
    "I don’t want to die, but I sure wouldn’t mind checking out for a while.",
    "My friend started a suicide hotline. Now he spends all his time listening to himself.",
    "I was going to tell a time-travel joke, but you didn’t like it.",
    "I walked into a bar... and ended up hitting rock bottom.",
    "I have a split personality... but that's okay because at least I get two opinions on my failures.",
    "I used to be afraid of dying alone. Now I’m just afraid of dying at all.",
    "My life’s not going well, but at least my funeral plans are looking great.",
    "I used to be a magician, but I quit. The audience couldn’t stop asking, 'Where’d you go?'",
    "I’m not saying my love life is bad, but even my imaginary girlfriend broke up with me."
];

const programmingJokes = [
    { text: "Why do programmers prefer dark mode? Because light attracts bugs!", tags: ['programming', 'software'] },
    { text: "!false - It's funny because it's true!", tags: ['programming', 'logic'] },
    { text: "Why do Java developers wear glasses? Because they don’t C#!", tags: ['java', 'csharp'] },
    { text: "There are only 10 types of people in the world: those who understand binary and those who don’t.", tags: ['binary', 'logic'] },
    { text: "How do you comfort a JavaScript bug? You console it.", tags: ['javascript', 'debugging'] },
    { text: "Why was the JavaScript developer sad? Because he didn’t ‘null’ his feelings.", tags: ['javascript', 'null'] },
    { text: "Why do programmers hate nature? It has too many bugs.", tags: ['bugs', 'programming'] },
    { text: "Why do Python developers hate Java? Because they can’t handle the Java beans!", tags: ['python', 'java'] },
    { text: "Why did the database administrator break up with the developer? He had too many commitments.", tags: ['database', 'sql'] },
    { text: "Why was the function so self-conscious? It had too many callbacks!", tags: ['javascript', 'callbacks'] },
    { text: "Why do C++ programmers need glasses? Because they don't see sharp.", tags: ['csharp', 'cplusplus'] },
    { text: "A programmer’s wife tells him: 'Go to the store and get a loaf of bread. If they have eggs, get a dozen.' He returns with 12 loaves of bread.", tags: ['logic', 'joke'] },
    { text: "Why did the programmer quit his job? He didn’t get arrays!", tags: ['arrays', 'programming'] },
    { text: "Why do programmers love Unix? Because it’s user-friendly, just very picky about its friends.", tags: ['unix', 'linux'] },
    { text: "Why do Java programmers wear headphones? Because they don't like threads!", tags: ['java', 'threads'] },
    { text: "Why do web developers hate the countryside? Too many cache cows.", tags: ['web', 'cache'] },
    { text: "Why do programmers always mix up Halloween and Christmas? Because Oct 31 == Dec 25.", tags: ['octal', 'binary'] },
    { text: "Why was the developer unhappy at his job? He wanted arrays.", tags: ['arrays', 'developer'] },
    { text: "Why don’t programmers like to go outside? The sunlight causes too many glares on their screens.", tags: ['programming', 'indoors'] },
    { text: "Why did the CSS developer go to therapy? Because he had too many unresolved issues.", tags: ['css', 'web'] },
    { text: "Why do functional programmers hate airports? Because they can’t handle side effects!", tags: ['functional', 'sideeffects'] },
    { text: "Why did the JavaScript developer go broke? Because he kept using ‘var’ instead of ‘let’!", tags: ['javascript', 'scoping'] },
    { text: "What’s the object-oriented way to become wealthy? Inheritance!", tags: ['oop', 'inheritance'] },
    { text: "Why do programmers prefer iOS development? Because Swift is a breeze!", tags: ['swift', 'ios'] },
    { text: "Why was the developer afraid of using Git? He didn’t want to commit!", tags: ['git', 'versioncontrol'] },
    { text: "How do you explain recursion to someone? Well, let me explain recursion to you…", tags: ['recursion', 'programming'] },
    { text: "Why was the JavaScript file sad? It was full of ‘undefined’ values.", tags: ['javascript', 'undefined'] },
    { text: "Why did the frontend developer break up with the backend developer? They had too many conflicts.", tags: ['frontend', 'backend'] },
    { text: "Why did the computer catch a cold? Because it left its Windows open.", tags: ['windows', 'software'] },
    { text: "Why do software engineers prefer to work at night? Because debugging is easier in the dark!", tags: ['debugging', 'coding'] },
    { text: "Why did the Git repository break up? It had too many issues.", tags: ['git', 'repository'] },
    { text: "Why do programmers hate SQL? Because they always have commitment issues!", tags: ['sql', 'database'] },
    { text: "Why don’t JavaScript developers trust each other? Because they always leave things undefined.", tags: ['javascript', 'trustissues'] },
    { text: "Why did the Java developer get locked out? He forgot his keys.", tags: ['java', 'keys'] },
    { text: "Why did the computer go to therapy? It had too many memory leaks.", tags: ['memory', 'ram'] },
    { text: "Why did the software engineer go broke? He used up all his cache.", tags: ['cache', 'money'] },
    { text: "What do you call a lazy programmer? A ‘while’ loop that never ends.", tags: ['loops', 'lazy'] },
    { text: "Why do web developers hate speed limits? Because they always want to go faster!", tags: ['web', 'performance'] },
    { text: "Why did the CSS developer go to the gym? He wanted to flexbox!", tags: ['css', 'flexbox'] },
    { text: "Why did the algorithm get promoted? It had a great runtime!", tags: ['algorithm', 'runtime'] },
    { text: "Why did the array feel empty? Because it had no elements left.", tags: ['array', 'data'] },
    { text: "Why don’t programmers use Facebook? They prefer GitHub for their commits.", tags: ['github', 'commits'] },
    { text: "What’s a programmer’s favorite game? Cache me if you can!", tags: ['cache', 'performance'] },
    { text: "Why did the programmer always carry a map? Because he kept losing his pointers!", tags: ['pointers', 'memory'] },
    { text: "Why was the JavaScript function sad? It didn’t get any arguments.", tags: ['javascript', 'functions'] },
    { text: "Why do developers never get lost? They follow the stack trace.", tags: ['debugging', 'stacktrace'] },
    { text: "Why was the cloud server so relaxed? It just went with the flow.", tags: ['cloud', 'server'] },
    { text: "Why do Python programmers love coffee? Because they don’t like Java.", tags: ['python', 'coffee'] },
    { text: "What’s a computer’s favorite snack? Microchips!", tags: ['hardware', 'chips'] }
];


const punsJokes = [
    { text: "What did the grape say when it got stepped on? Nothing - it just let out a little wine!", tags: ['food', 'drink'] },
    { text: "Time flies like an arrow. Fruit flies like a banana.", tags: ['food', 'wordplay'] },
    { text: "I told my wife she should embrace her mistakes. She gave me a hug.", tags: ['relationship', 'wordplay'] },
    { text: "I’m reading a book about anti-gravity. It’s impossible to put down!", tags: ['science', 'books'] },
    { text: "What’s the best way to watch a fly fishing tournament? Live stream it.", tags: ['sports', 'wordplay'] },
    { text: "Why don’t skeletons fight each other? They don’t have the guts.", tags: ['halloween', 'wordplay'] },
    { text: "I used to play piano by ear, but now I use my hands.", tags: ['music', 'wordplay'] },
    { text: "I wondered why the baseball kept getting bigger. Then it hit me.", tags: ['sports', 'wordplay'] },
    { text: "Did you hear about the cheese factory explosion? There was nothing left but de-brie.", tags: ['food', 'cheese'] },
    { text: "Why don’t some couples go to the gym? Because some relationships don’t work out.", tags: ['fitness', 'relationship'] },
    { text: "I used to be a baker, but I couldn’t make enough dough.", tags: ['food', 'career'] },
    { text: "I told my suitcase that there will be no vacations this year. Now I’m dealing with emotional baggage.", tags: ['travel', 'wordplay'] },
    { text: "I tried to make a belt out of watches. It was a waist of time.", tags: ['fashion', 'wordplay'] },
    { text: "What do you call fake spaghetti? An impasta!", tags: ['food', 'pasta'] },
    { text: "The shovel was a ground-breaking invention.", tags: ['tools', 'wordplay'] },
    { text: "I don’t trust stairs. They’re always up to something.", tags: ['household', 'wordplay'] },
    { text: "Why did the scarecrow win an award? Because he was outstanding in his field!", tags: ['farm', 'wordplay'] },
    { text: "I’m friends with all electricians. We have such good current connections.", tags: ['electricity', 'friendship'] },
    { text: "What did the janitor say when he jumped out of the closet? Supplies!", tags: ['work', 'wordplay'] },
    { text: "I couldn’t figure out why I don’t like elevator music. Then it hit me — it’s bad on so many levels.", tags: ['music', 'wordplay'] },
    { text: "I told my friend she should do lunges to stay in shape. That was a big step forward.", tags: ['fitness', 'wordplay'] },
    { text: "I don't play soccer because I enjoy the sport. I'm just doing it for kicks.", tags: ['sports', 'wordplay'] },
    { text: "Why do cows have hooves instead of feet? Because they lactose.", tags: ['animals', 'wordplay'] },
    { text: "Why was the math book sad? It had too many problems.", tags: ['math', 'school'] },
    { text: "I got a job at a bakery because I kneaded dough.", tags: ['food', 'career'] },
    { text: "The bicycle couldn’t stand up by itself—it was two-tired!", tags: ['transportation', 'wordplay'] },
    { text: "Why don’t eggs tell jokes? Because they might crack up.", tags: ['food', 'eggs'] },
    { text: "I stayed up all night wondering where the sun went. Then it dawned on me.", tags: ['nature', 'wordplay'] },
    { text: "I used to be a banker, but I lost interest.", tags: ['career', 'money'] },
    { text: "Why did the tomato blush? Because it saw the salad dressing.", tags: ['food', 'wordplay'] },
    { text: "I told my dog a joke, but he didn’t laugh. Guess he’s a little ruff.", tags: ['animals', 'dogs'] },
    { text: "Why do birds fly south for the winter? Because it’s too far to walk!", tags: ['nature', 'birds'] },
    { text: "I opened a bakery for clumsy people. It’s called ‘Oops, I Crumbed Again.’", tags: ['food', 'bakery'] },
    { text: "I told my plants a joke. Now they’re rooting for me!", tags: ['nature', 'plants'] },
    { text: "I quit my job at the shoe factory. It just wasn’t my sole purpose.", tags: ['career', 'shoes'] },
    { text: "I made a pencil with two erasers. It was pointless.", tags: ['school', 'wordplay'] },
    { text: "What did the big flower say to the little flower? ‘Hey, bud!’", tags: ['nature', 'flowers'] },
    { text: "I wanted to learn how to make ice cream, but I just couldn’t churn up the motivation.", tags: ['food', 'wordplay'] },
    { text: "I’m on a seafood diet. I see food, and I eat it.", tags: ['food', 'diet'] },
    { text: "Why did the golfer bring an extra pair of pants? In case he got a hole in one.", tags: ['sports', 'golf'] },
    { text: "I asked my dog what’s two minus two. He said nothing.", tags: ['animals', 'wordplay'] },
    { text: "Why did the music teacher go to jail? Because she got caught with too many notes.", tags: ['music', 'jail'] },
    { text: "What do you call an alligator in a vest? An in-vest-igator!", tags: ['animals', 'wordplay'] },
    { text: "I was struggling to figure out how lightning works. Then it struck me.", tags: ['science', 'wordplay'] },
    { text: "Why don’t vampires like garlic? Because it makes them coffin.", tags: ['halloween', 'wordplay'] },
    { text: "I lost my mood ring… I don’t know how I feel about it.", tags: ['fashion', 'wordplay'] },
    { text: "I told my friend 10 jokes to make him laugh. Sadly, no pun in ten did.", tags: ['wordplay', 'jokes'] },
    { text: "I thought about becoming a baker, but I couldn’t handle the knead for attention.", tags: ['food', 'career'] },
    { text: "I got a job at a calendar factory, but I got fired for taking too many days off.", tags: ['career', 'time'] }
];


class JokeManager {
    constructor() {
        this.jokes = {
            dad: dadJokes.map(joke => ({ text: joke, rating: 0, votes: 0, tags: ['dad'], language: 'en' })),
            dark: darkJokes.map(joke => ({ text: joke, rating: 0, votes: 0, tags: ['dark'], language: 'en' })),
            programming: programmingJokes.map(joke => ({ ...joke, rating: 0, votes: 0, language: 'en' })),
            puns: punsJokes.map(joke => ({ ...joke, rating: 0, votes: 0, language: 'en' }))
        };
        this.cache = new Map();
    }

    rateJoke(type, index, rating) {
        if (rating < 1 || rating > 5) throw new Error("Rating must be between 1 and 5");
        const joke = this.jokes[type][index];
        joke.rating = ((joke.rating * joke.votes) + rating) / (joke.votes + 1);
        joke.votes++;
        return joke.rating;
    }

    getJokesByTag(tag) {
        const cacheKey = `tag_${tag}`;
        if (this.cache.has(cacheKey)) return this.cache.get(cacheKey);

        const filtered = Object.values(this.jokes)
            .flat()
            .filter(joke => joke.tags.includes(tag));
        
        this.cache.set(cacheKey, filtered);
        return filtered;
    }

    getTopRated(type, limit = 10) {
        return this.jokes[type]
            .sort((a, b) => b.rating - a.rating)
            .slice(0, limit);
    }

    submitJoke(joke, type, tags = [], language = 'en') {
        const newJoke = {
            text: joke,
            rating: 0,
            votes: 0,
            tags,
            language
        };
        this.jokes[type].push(newJoke);
        this.clearCache();
        return newJoke;
    }

    clearCache() {
        this.cache.clear();
    }
}

const jokeManager = new JokeManager();

function generateJokes(jokeArray, amount = 1, numbered = false) {
    if (amount < 1) {
        throw new Error("The minimum amount of jokes is 1.");
    }
    if (amount > jokeArray.length) {
        throw new Error(`Cannot return ${amount} jokes, only ${jokeArray.length} available.`);
    }

    if (amount === 1) {
        return numbered ? `1. ${jokeArray[Math.floor(Math.random() * jokeArray.length)]}` 
                      : jokeArray[Math.floor(Math.random() * jokeArray.length)];
    }

    const selectedJokes = new Set();
    while (selectedJokes.size < amount) {
        const randomJoke = jokeArray[Math.floor(Math.random() * jokeArray.length)];
        selectedJokes.add(randomJoke);
    }

    const jokesArray = Array.from(selectedJokes);
    if (numbered) {
        return jokesArray.map((joke, index) => `${index + 1}. ${joke}`).join('\n');
    }

    return Object.assign({}, ...jokesArray.map((joke, index) => ({ [index + 1]: joke })));
}

const joke = (amount = 1, numbered = false) => {
    throw new Error("Please specify joke type: joke.dad(), joke.dark(), joke.programming(), or joke.puns()");
};

joke.dad = (amount = 1, numbered = false, minRating = 0) => {
    const filtered = jokeManager.jokes.dad.filter(j => j.rating >= minRating);
    return generateJokes(filtered.map(j => j.text), amount, numbered);
};

joke.dark = (amount = 1, numbered = false, minRating = 0) => {
    const filtered = jokeManager.jokes.dark.filter(j => j.rating >= minRating);
    return generateJokes(filtered.map(j => j.text), amount, numbered);
};

joke.programming = (amount = 1, numbered = false, minRating = 0) => {
    const filtered = jokeManager.jokes.programming.filter(j => j.rating >= minRating);
    return generateJokes(filtered.map(j => j.text), amount, numbered);
};

joke.puns = (amount = 1, numbered = false, minRating = 0) => {
    const filtered = jokeManager.jokes.puns.filter(j => j.rating >= minRating);
    return generateJokes(filtered.map(j => j.text), amount, numbered);
};

joke.custom = function(jokes, amount = 1, numbered = false) {
    if (!jokes) {
        throw new Error("No jokes provided");
    }
    const jokeArray = Array.isArray(jokes) ? jokes : [jokes];
    return generateJokes(jokeArray, amount, numbered);
}

joke.generate = function({ type, numbered = false, jokes = 1 }) {
    if (type === 'dad') {
        return joke.dad(jokes, numbered);
    } else if (type === 'dark') {
        return joke.dark(jokes, numbered);
    } else if (type === 'programming') {
        return joke.programming(jokes, numbered);
    } else if (type === 'puns') {
        return joke.puns(jokes, numbered);
    } else {
        throw new Error("Invalid joke type. Please specify 'dad', 'dark', 'programming', or 'puns'.");
    }
};

joke.submitJoke = (text, type, tags, language) => jokeManager.submitJoke(text, type, tags, language);
joke.rateJoke = (type, index, rating) => jokeManager.rateJoke(type, index, rating);
joke.getTopRated = (type, limit) => jokeManager.getTopRated(type, limit);
joke.getByTag = (tag) => jokeManager.getJokesByTag(tag);

module.exports = { joke };
